pkg/controller/controller.go

```diff
@@ -602,10 +602,10 @@ func (p *csiProvisioner) prepareProvision(ctx context.Context, claim *v1.Persist
 		return nil, controller.ProvisioningFinished, err
 	}
 
-        parametersCopy := make(map[string]string)
-        for k, v := range sc.Parameters {
-                parametersCopy[k] = v
-        }
+	parametersCopy := make(map[string]string)
+	for k, v := range sc.Parameters {
+		parametersCopy[k] = v
+	}
```

Comment:
Since line 698/693 overwrites the req.Paramenters anyway, can just leave req.Parameters = nil on line 612 and populate it on line 698/693? That will probably avoid such mistake in future.

Reply:
Also, does this mean that the fix that made the parametersCopy is actually not fixing anything?

Reply:
> Since line 698/693 overwrites the req.Paramenters anyway, can just leave req.Parameters = nil on line 612 and populate it on line 698/693? That will probably avoid such mistake in future.

That could be possible, but I mainly try to minimize the changes to this codebase as it's a fork of the open source one. I moved the parameters addition to be right before create so we don't see this again. Also, I'm trying to move away from this fork soon so we don't need to maintain it.

Re: this is a good point, but I just checked and `removePrefixedParameters`, where we re-assign sc.Parameters, actually creates a new map anyway. So there will not be a concurrent write issue there.

Reply:
> Re: this is a good point, but I just checked and removePrefixedParameters, where we re-assign sc.Parameters, actually creates a new map anyway. So there will not be a concurrent write issue there.

Right, so we should probably revert the fix that added parametersCopy  since it is a dead code, basically.

Reply:
You are right Neelesh, the `removePrefixedParameters` does make a copy, so there is something else which is causing the concurrent map read-write access.

Reply:
`removePrefixedParameters` makes a copy of sc.Parameters later in prepareProvision after we were writing to sc.Parameters. A concurrent read/write was still possible here as we write to the map before this copy is made.

 With @adityadani's fix, we create a copy of sc.Paramters up front before writing to it in `p.addPVCMetadataParams`, which is why the fix is still valid.

Reply:
Yes, it makes writing to the copy "safe" but anything you write to the copy is discarded by the `removePrefixedParameters` making the "safety" misleading and pointless. So, I still think this confusion should be removed by just setting the req.Parameters to nil in the beginning.

Reply:
Agreed, that is misleading. If this was code we owned entirely I would agree to assign as nil initially.

However, since we're always pulling changes from the upstream CSI version of this repo, I'm afraid setting req.Parameters as nil could have unintended consequences of its own (accessing nil map, assigning to nil map, etc).

As a result, I would prefer to minimize changes we make to this repo in order for our changes to integrate reliability with the upstream version. An upstream change could easily cause a panic if req.Parameters is nil.

Reply:
Fair enough. But please consider removing our change that makes a copy of the params. IMHO, it adds to the confusion without adding any value.